pkg/rbd/controllerserver.go

```diff
@@ -485,25 +512,46 @@ func (cs *ControllerServer) validateSnapshotReq(req *csi.CreateSnapshotRequest)
 	return nil
 }
 
-func (cs *ControllerServer) doSnapshot(rbdSnap *rbdSnapshot, secret map[string]string) error {
-	err := createSnapshot(rbdSnap, rbdSnap.AdminID, secret)
+func (cs *ControllerServer) doSnapshot(snapName, snapID, VolID string, options, secret map[string]string) (*rbdSnapshot, error) {
+	rbdSnap, err := getRBDSnapshotOptions(options)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	uniqueID := uuid.NewUUID().String()
+	rbdVolume, err := getRBDVolumeByID(VolID)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "Source Volume ID %s cannot found", VolID)
+	}
+	if !hasSnapshotFeature(rbdVolume.ImageFeatures) {
+		return nil, status.Errorf(codes.InvalidArgument, "volume(%s) has not snapshot feature(layering)", VolID)
+	}
+	rbdSnap.VolName = rbdVolume.VolName
+	rbdSnap.SnapName = snapName
+	if snapID == "" {
+		snapID = "csi-rbd-" + rbdVolume.VolName + "-snap-" + uniqueID
```

Comment:
can this be made a helper function like `createSnapId(..)` so that the format is sortof documented somewhere

Maybe a `parseSnapId(..)` function makes sense too.

Reply:
yeah, might be good to move that out even if we don't currently use it directly from anywhere else.  It might be worth considering generalizing it a bit and see where we can share it.  I'd also consider setting up some constants rather than raw strings in here.  I'm also not sure about the looooong naming convention, but don't have a better idea off the top of my head :)